### ISSUE TITLE
How does custom data training process actually work?

### ISSUE THREAD

**Issue by sardosan** [ultralytics/ultralytics#6255]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

first of all thank you for your great work.

I try to understand the functioning of **model.train and model.val** within every kind of website and issue but it seems that is not clear as well as clarified. I need to understand how the training process actually works.

In the following I would describe my situation.

I have a custom dataset of different .jpeg images and corresponding .txt files with the correct yolo annotation, respectively distributed in images and labels directories..
The **dataset structure** is as the one provided below:

C:.
+---data
    +---test
    ¦   +---images
    ¦   +---labels
    +---train
    ¦   +---images
    ¦   +---labels
    +---val
        +---images
        +---labels

#########################################

The **data.yaml** file is as the one provided below:

`path: C:/Users/matte/Desktop/data
train: images/train
val: images/val
test: images/test

nc: 31
names: ['Label', 'PlasticPart', 'ClampPart', 'Nut', 'BoltNutSet', 'Cutter', 'Rock', 'FuelCap', 'AdjustableClamp', 'LuggagePart', 'Wire', 'Hammer', 'SodaCan', 'MetalPart', 'AdjustableWrench', 'Pliers', 'BoltWasher', 'Wrench', 'MetalSheet', 'Screw', 'Battery', 'Screwdriver', 'Washer', 'Tape', 'LuggageTag', 'Bolt', 'Wood', 'Hose', 'PaintChip', 'Nail', 'Pen']`

#########################################

The code for **training** the model is as the one provided below:

`from ultralytics import YOLO
import torch

def train_model():
    model = YOLO("yolov8n.pt")
    results = model.train(data="data.yaml", epochs=1, batch=32) 

if __name__ == '__main__':
    train_model()`

#########################################

The code for **validating** the model is as the one provided below:

`from ultralytics import YOLO

if __name__ == '__main__':
    model = YOLO(r"C:\Users\matte\Documents\GitHub\ultralytics\runs\detect\train\weights\best.pt") 

    metrics = model.val()`

#########################################

The code for **using** the model with my integrated webcam in realtime is as the one provided below:

`from ultralytics import YOLO

model = YOLO(r"C:\Users\matte\Documents\GitHub\ultralytics\runs\detect\train\weights\best.pt") # path to your best model under run/detect/train folder

results = model.predict(source="0",show=True)

print (results)`

#########################################

I want to know how .train and .val consider the data.yaml file.
In particular:
1. Does **.train** function train the model on the train images contained in the _train declared path into the data.yaml_ file and then fine change/tune the model weights (i.e. increase the model performance) through a validation on the val images contained in the _val declared path into the data.yaml_ file?
2. Does **.val** function evaluate the model on the test images contained in the _test declared path into the data.yaml_ file making a prediction on those images and comparing the predicted results with the corresponding real one.
3. What is the best procedure/pipeline suggested to train a yolov8 model? Is it to split the entire dataset into train/val/test? 
4. Are the structure of the custom dataset and the data.yaml file information correctly configurated in my case?

Thank you in advance for your answers.

**Comment by sardosan** [comment]
Hi @glenn-jocher and nice to e-meet you.

Thank you for your great work and answer.

I think that all is clear apart from the 'test' images directory and its usage. 
How can I use that information? Why should I declare (consequence of the three-partiotion splitting) a 'test' set? What is it used for?

Is correct I can use it through the '.val' function? If yes, how can I do it?

Would you be so kind to elaborate on this?

Thank you!

**Comment by glenn-jocher** [comment]
Nice to e-meet you too, @sardosan!

Your query about the usage of the 'test' images directory is a common one, so let's clarify that:

The 'test' dataset is intended to be a set of data that the model has never seen during the training or validation process. It serves as the best representation of how the model will perform in real-world scenarios or on completely unseen data. It's a final step to ensure that model performance metrics are not biased by any data the model has been exposed to during training or validation.

You can indeed use the `.val` function to validate your model on the 'test' dataset. To do this, you would adjust your code to specify the 'test' set as follows:

```
metrics = model.val(data='data.yaml', task='test')
```

This command tells the `.val` method to use the data from the 'test' portion specified in your `data.yaml`.

It is crucial to note that evaluating the model on the 'test' dataset should ideally happen after you have finished all model tuning and are ready to assess final performance metrics.

For questions specific to the YOLOv8 implementation methods, the best source is the Ultralytics Documentation, which we maintain with detailed examples, explanations, and use-case scenarios to help guide users through various processes, including training, validation, and testing.

The documentation covers many questions related to custom training, data preparation, and best practices for YOLOv8. It is kept up to date with the latest changes and features of the models and methods.

Thank you for reaching out, and I wish you the best with your YOLOv8 model testing and deployment!
